File: src/main/java/io/polygloat/model/UserAccount.java
package io.polygloat.model;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"username"}, name = "useraccount_username"),
        @UniqueConstraint(columnNames = {"third_party_auth_type", "third_party_auth_id"}, name = "useraccount_authtype_auth_id"),

})
public class UserAccount extends AuditModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String username;

    private String password;

    private String name;

    @OneToMany(mappedBy = "createdBy")
    private Set<Repository> createdRepositories = new HashSet<>();

    @OneToMany(mappedBy = "user")
    private Set<Permission> permissions;

    @Enumerated(EnumType.STRING)
    private Role role = Role.USER;

    @Column(name = "third_party_auth_type")
    private String thirdPartyAuthType;

    @Column(name = "third_party_auth_id")
    private String thirdPartyAuthId;

    @Column(name = "reset_password_code")
    private String resetPasswordCode;


    public UserAccount() {
    }

    public UserAccount(String username) {
        this.username = username;
    }

    public UserAccount(Long id, String username, String password, String name, Set<Repository> createdRepositories, Set<Permission> permissions, Role role, String thirdPartyAuthType, String thirdPartyAuthId, String resetPasswordCode) {
        this.id = id;
        this.username = username;
        this.password = password;
        this.name = name;
        this.createdRepositories = createdRepositories;
        this.permissions = permissions;
        this.role = role;
        this.thirdPartyAuthType = thirdPartyAuthType;
        this.thirdPartyAuthId = thirdPartyAuthId;
        this.resetPasswordCode = resetPasswordCode;
    }

    private static Set<Repository> $default$createdRepositories() {
        return new HashSet<>();
    }

    private static Role $default$role() {
        return Role.USER;
    }

    public static UserAccountBuilder builder() {
        return new UserAccountBuilder();
    }

    public Long getId() {
        return this.id;
    }

    public String getUsername() {
        return this.username;
    }

    public String getPassword() {
        return this.password;
    }

    public String getName() {
        return this.name;
    }

    public Set<Repository> getCreatedRepositories() {
        return this.createdRepositories;
    }

    public Set<Permission> getPermissions() {
        return this.permissions;
    }

    public Role getRole() {
        return this.role;
    }

    public String getThirdPartyAuthType() {
        return this.thirdPartyAuthType;
    }

    public String getThirdPartyAuthId() {
        return this.thirdPartyAuthId;
    }

    public String getResetPasswordCode() {
        return this.resetPasswordCode;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setCreatedRepositories(Set<Repository> createdRepositories) {
        this.createdRepositories = createdRepositories;
    }

    public void setPermissions(Set<Permission> permissions) {
        this.permissions = permissions;
    }

    public void setRole(Role role) {
        this.role = role;
    }

    public void setThirdPartyAuthType(String thirdPartyAuthType) {
        this.thirdPartyAuthType = thirdPartyAuthType;
    }

    public void setThirdPartyAuthId(String thirdPartyAuthId) {
        this.thirdPartyAuthId = thirdPartyAuthId;
    }

    public void setResetPasswordCode(String resetPasswordCode) {
        this.resetPasswordCode = resetPasswordCode;
    }

    public boolean equals(final Object o) {
        if (o == this) return true;
        if (!(o instanceof UserAccount)) return false;
        final UserAccount other = (UserAccount) o;
        if (!other.canEqual((Object) this)) return false;
        if (!super.equals(o)) return false;
        final Object this$id = this.getId();
        final Object other$id = other.getId();
        if (this$id == null ? other$id != null : !this$id.equals(other$id)) return false;
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof UserAccount;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = super.hashCode();
        final Object $id = this.getId();
        result = result * PRIME + ($id == null ? 43 : $id.hashCode());
        return result;
    }

    public String toString() {
        return "UserAccount(id=" + this.getId() + ", username=" + this.getUsername() + ", name=" + this.getName() + ")";
    }

    public enum Role {
        USER,
        ADMIN
    }

    public static class UserAccountBuilder {
        private Long id;
        private String username;
        private String password;
        private String name;
        private Set<Repository> createdRepositories$value;
        private boolean createdRepositories$set;
        private Set<Permission> permissions;
        private Role role$value;
        private boolean role$set;
        private String thirdPartyAuthType;
        private String thirdPartyAuthId;
        private String resetPasswordCode;

        UserAccountBuilder() {
        }

        public UserAccount.UserAccountBuilder id(Long id) {
            this.id = id;
            return this;
        }

        public UserAccount.UserAccountBuilder username(String username) {
            this.username = username;
            return this;
        }

        public UserAccount.UserAccountBuilder password(String password) {
            this.password = password;
            return this;
        }

        public UserAccount.UserAccountBuilder name(String name) {
            this.name = name;
            return this;
        }

        public UserAccount.UserAccountBuilder createdRepositories(Set<Repository> createdRepositories) {
            this.createdRepositories$value = createdRepositories;
            this.createdRepositories$set = true;
            return this;
        }

        public UserAccount.UserAccountBuilder permissions(Set<Permission> permissions) {
            this.permissions = permissions;
            return this;
        }

        public UserAccount.UserAccountBuilder role(Role role) {
            this.role$value = role;
            this.role$set = true;
            return this;
        }

        public UserAccount.UserAccountBuilder thirdPartyAuthType(String thirdPartyAuthType) {
            this.thirdPartyAuthType = thirdPartyAuthType;
            return this;
        }

        public UserAccount.UserAccountBuilder thirdPartyAuthId(String thirdPartyAuthId) {
            this.thirdPartyAuthId = thirdPartyAuthId;
            return this;
        }

        public UserAccount.UserAccountBuilder resetPasswordCode(String resetPasswordCode) {
            this.resetPasswordCode = resetPasswordCode;
            return this;
        }

        public UserAccount build() {
            Set<Repository> createdRepositories$value = this.createdRepositories$value;
            if (!this.createdRepositories$set) {
                createdRepositories$value = UserAccount.$default$createdRepositories();
            }
            Role role$value = this.role$value;
            if (!this.role$set) {
                role$value = UserAccount.$default$role();
            }
            return new UserAccount(id, username, password, name, createdRepositories$value, permissions, role$value, thirdPartyAuthType, thirdPartyAuthId, resetPasswordCode);
        }

        public String toString() {
            return "UserAccount.UserAccountBuilder(id=" + this.id + ", username=" + this.username + ", password=" + this.password + ", name=" + this.name + ", createdRepositories$value=" + this.createdRepositories$value + ", permissions=" + this.permissions + ", role$value=" + this.role$value + ", thirdPartyAuthType=" + this.thirdPartyAuthType + ", thirdPartyAuthId=" + this.thirdPartyAuthId + ", resetPasswordCode=" + this.resetPasswordCode + ")";
        }
    }
}

File: backend/data/src/main/kotlin/io/tolgee/model/UserAccount.kt
package io.tolgee.model

import io.hypersistence.utils.hibernate.type.array.ListArrayType
import io.tolgee.activity.annotation.ActivityLoggedEntity
import io.tolgee.api.IUserAccount
import io.tolgee.component.ThirdPartyAuthTypeConverter
import io.tolgee.model.enums.ThirdPartyAuthType
import io.tolgee.model.slackIntegration.SlackConfig
import io.tolgee.model.slackIntegration.SlackUserConnection
import io.tolgee.model.task.Task
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.ManyToMany
import jakarta.persistence.OneToMany
import jakarta.persistence.OneToOne
import jakarta.persistence.OrderBy
import jakarta.persistence.Transient
import jakarta.validation.constraints.NotBlank
import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.Type
import java.util.Date

@Entity
@ActivityLoggedEntity
data class UserAccount(
  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  override var id: Long = 0L,
  @field:NotBlank
  override var username: String = "",
  var password: String? = null,
  var name: String = "",
  @Enumerated(EnumType.STRING)
  var role: Role? = Role.USER,
  /**
   * This property is redundant - it's value can be derived from other existing properties.
   * Kept for legacy reasons.
   *
   * It's value follows these rules, but there are some edge cases related to old accounts:
   * - NATIVE ->      password != null && thirdPartyAuthType == GITHUB | GOOGLE | OAUTH | null
   * - THIRD_PARTY -> password == null && thirdPartyAuthType == GITHUB | GOOGLE | OAUTH
   * - MANAGED ->     password == null && thirdPartyAuthType == SSO | SSO_GLOBAL
   */
  @Enumerated(EnumType.STRING)
  @Column(name = "account_type")
  override var accountType: AccountType? = AccountType.LOCAL,
) : AuditModel(),
  ModelWithAvatar,
  IUserAccount,
  EntityWithId {
  @Column(name = "totp_key", columnDefinition = "bytea")
  override var totpKey: ByteArray? = null

  @Type(ListArrayType::class)
  @Column(name = "mfa_recovery_codes", columnDefinition = "text[]")
  var mfaRecoveryCodes: List<String> = emptyList()

  @Column(name = "tokens_valid_not_before")
  var tokensValidNotBefore: Date? = null

  @OneToMany(mappedBy = "user", orphanRemoval = true)
  var permissions: MutableSet<Permission> = mutableSetOf()

  @OneToOne(mappedBy = "userAccount", fetch = FetchType.LAZY, optional = true)
  var emailVerification: EmailVerification? = null

  @OneToOne(
    mappedBy = "userAccount",
    fetch = FetchType.LAZY,
    cascade = [CascadeType.REMOVE],
    orphanRemoval = true,
    optional = true,
  )
  var authProviderChangeRequest: AuthProviderChangeRequest? = null

  @Column(name = "third_party_auth_type")
  @Convert(converter = ThirdPartyAuthTypeConverter::class)
  var thirdPartyAuthType: ThirdPartyAuthType? = null

  @Column(name = "sso_refresh_token", columnDefinition = "TEXT")
  var ssoRefreshToken: String? = null

  @Column(name = "sso_session_expiry")
  var ssoSessionExpiry: Date? = null

  @Column(name = "third_party_auth_id")
  var thirdPartyAuthId: String? = null

  @Column(name = "reset_password_code")
  var resetPasswordCode: String? = null

  @OrderBy("id ASC")
  @OneToMany(mappedBy = "user", orphanRemoval = true)
  var organizationRoles: MutableList<OrganizationRole> = mutableListOf()

  @OneToOne(mappedBy = "userAccount", fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE], orphanRemoval = true)
  var preferences: UserPreferences? = null

  @OneToMany(mappedBy = "userAccount", orphanRemoval = true)
  var pats: MutableList<Pat>? = mutableListOf()

  @OneToMany(mappedBy = "userAccount", orphanRemoval = true)
  var apiKeys: MutableList<ApiKey>? = mutableListOf()

  @OneToMany(mappedBy = "createdBy", orphanRemoval = true)
  var invitations: MutableList<Invitation>? = mutableListOf()

  override var avatarHash: String? = null

  @Column(name = "deleted_at")
  var deletedAt: Date? = null

  @Column(name = "disabled_at")
  var disabledAt: Date? = null

  @Column(name = "is_initial_user", nullable = false)
  @ColumnDefault("false")
  override var isInitialUser: Boolean = false

  @Column(name = "password_changed", nullable = false)
  @ColumnDefault("true")
  var passwordChanged: Boolean = true

  /**
   * Whether user is created only to be used as a part of demo data
   */
  @ColumnDefault("false")
  var isDemo: Boolean = false

  @OneToMany(mappedBy = "userAccount", fetch = FetchType.LAZY, orphanRemoval = true)
  var slackUserConnection: MutableList<SlackUserConnection> = mutableListOf()

  @OneToMany(mappedBy = "userAccount", fetch = FetchType.LAZY, orphanRemoval = true)
  var slackConfig: MutableList<SlackConfig> = mutableListOf()

  @ManyToMany(mappedBy = "assignees")
  var tasks: MutableSet<Task> = mutableSetOf()

  constructor(
    id: Long?,
    username: String?,
    password: String?,
    name: String?,
    permissions: MutableSet<Permission>,
    role: Role = Role.USER,
    accountType: AccountType = AccountType.LOCAL,
    thirdPartyAuthType: ThirdPartyAuthType?,
    thirdPartyAuthId: String?,
    resetPasswordCode: String?,
  ) : this(id = 0L, username = "", password, name = "") {
    this.permissions = permissions
    this.role = role
    this.accountType = accountType
    this.thirdPartyAuthType = thirdPartyAuthType
    this.thirdPartyAuthId = thirdPartyAuthId
    this.resetPasswordCode = resetPasswordCode
  }

  enum class Role {
    USER,
    ADMIN,
    SUPPORTER,
  }

  enum class AccountType {
    LOCAL,
    MANAGED,
    THIRD_PARTY,
  }

  @Transient
  override var disableActivityLogging: Boolean = false
}

fun UserAccount.isAdmin(): Boolean {
  return role == UserAccount.Role.ADMIN
}

fun UserAccount.isSupporter(): Boolean {
  return role == UserAccount.Role.SUPPORTER
}

fun UserAccount.isSupporterOrAdmin(): Boolean {
  return role == UserAccount.Role.SUPPORTER || role == UserAccount.Role.ADMIN
}
